File: python-gdb-activities-20/activity19/gdb/service/transfer_service.py
# gdb/service/transfer_service.py
from gdb.domain.iaccount import IAccount
from gdb.exceptions.account_exception import AccountException
from gdb.exceptions.inactive_account_exception import InactiveAccountException
from gdb.exceptions.insufficient_balance_exception import InsufficientBalanceException
from gdb.exceptions.invalid_pin_exception import InvalidPinException

class TransferService:
    """Service managing funds transfer operations, validations, and daily tier limits."""

    def __init__(self) -> None:
        pass

    def transfer(self, from_acc: IAccount, to_acc: IAccount, amount: float, pin: int) -> None:
        if from_acc is None or to_acc is None:
            raise AccountException("Source and destination accounts are required")
        if not from_acc.is_active() or not to_acc.is_active():
            raise InactiveAccountException("Both accounts must be active to transfer funds")
        if not from_acc.verify_pin(pin):
            raise InvalidPinException("Incorrect PIN")
        if not from_acc.can_withdraw(amount):
            raise InsufficientBalanceException(f"Insufficient balance for transfer of Rs. {amount}")
        if hasattr(from_acc, "reset_daily_transfer_if_needed"):
            from_acc.reset_daily_transfer_if_needed()
        if hasattr(from_acc, "can_transfer") and not from_acc.can_transfer(amount):
            remaining = from_acc.get_remaining_daily_transfer_limit() if hasattr(from_acc, "get_remaining_daily_transfer_limit") else 0.0
            raise AccountException(f"Daily transfer limit exceeded. Remaining today: Rs. {remaining}")

        from_acc.withdraw(amount, pin)
        to_acc.deposit(amount)
        if hasattr(from_acc, "update_daily_transfer_total"):
            from_acc.update_daily_transfer_total(amount)

File: java-gdb-activities-20/activity16-solution/src/com/gdb/service/TransferService.java
package com.gdb.service;

import com.gdb.domain.*;
import com.gdb.exceptions.*;
import java.time.LocalDateTime;

public class TransferService {

    public TransferService() {
    }

    public void transfer(IAccount from, IAccount to, double amount, int pin)
            throws AccountException {
        if (from == null || to == null) {
            throw new AccountException("Source and destination accounts are required");
        }
        if (!from.isActive() || !to.isActive()) {
            throw new InactiveAccountException("Both accounts must be active to transfer funds");
        }
        if (!from.verifyPin(pin)) {
            throw new InvalidPinException("Incorrect PIN");
        }
        if (!from.canWithdraw(amount)) {
            throw new InsufficientBalanceException("Insufficient balance for transfer of Rs. " + amount);
        }
        Account source = (Account) from;
        source.resetDailyTransferIfNeeded();
        if (!source.canTransfer(amount)) {
            throw new AccountException("Daily transfer limit exceeded. Remaining today: Rs. " + source.getRemainingDailyTransferLimit());
        }
        from.withdraw(amount, pin);
        to.deposit(amount);
        source.updateDailyTransferTotal(amount);
    }

    public Transaction transferWithTransaction(IAccount from, IAccount to, 
                                               double amount, int pin) throws AccountException {
        transfer(from, to, amount, pin);
        return new Transaction(
            Transaction.generateId(),
            LocalDateTime.now(),
            from.getAccountNumber(),
            TransactionType.TRANSFER,
            amount,
            from.getBalance(),
            "SUCCESS",
            "Transfer of Rs. " + amount + " to Account #" + to.getAccountNumber(),
            from.getAccountNumber(),
            to.getAccountNumber()
        );
    }
}
